hypothesis-python/src/hypothesis/internal/conjecture/data.py

```diff
@@ -856,31 +856,221 @@ def kill_branch(self) -> None:
         """Mark this part of the tree as not worth re-exploring."""
 
     def draw_integer(
-        self, value: int, *, was_forced: bool, kwargs: IntegerKWargs
+        self, value: int, *, kwargs: IntegerKWargs, was_forced: bool
     ) -> None:
         pass
 
     def draw_float(
-        self, value: float, *, was_forced: bool, kwargs: FloatKWargs
+        self, value: float, *, kwargs: FloatKWargs, was_forced: bool
     ) -> None:
         pass
 
     def draw_string(
-        self, value: str, *, was_forced: bool, kwargs: StringKWargs
+        self, value: str, *, kwargs: StringKWargs, was_forced: bool
     ) -> None:
         pass
 
     def draw_bytes(
-        self, value: bytes, *, was_forced: bool, kwargs: BytesKWargs
+        self, value: bytes, *, kwargs: BytesKWargs, was_forced: bool
     ) -> None:
         pass
 
     def draw_boolean(
-        self, value: bool, *, was_forced: bool, kwargs: BooleanKWargs
+        self, value: bool, *, kwargs: BooleanKWargs, was_forced: bool
     ) -> None:
         pass
 
 
+IRTreeNodeType: TypeAlias = Union["IRTreeNode", "IRTreeLeaf"]
+LocationType: TypeAlias = Tuple[int, int]
+
+
+@attr.s(slots=True)
+class IRTreeLeaf:
+    ir_type: IRTypeName = attr.ib()
+    value: IRType = attr.ib()
+    kwargs: IRKWargsType = attr.ib()
+    was_forced: bool = attr.ib()
+    depth: int = attr.ib()
+    index_in_parent: int = attr.ib()
+
+    location: LocationType = attr.ib(init=False)
+
+    def __attrs_post_init__(self):
+        self.location = (self.depth, self.index_in_parent)
+
+    def __repr__(self):
+        return f"{self.ir_type} {self.value} {self.kwargs}"
+
+    def copy(self, *, with_value: Optional[IRType] = None) -> "IRTreeLeaf":
+        value = self.value if with_value is None else with_value
+        return IRTreeLeaf(
+            ir_type=self.ir_type,
+            value=value,
+            kwargs=self.kwargs,
+            was_forced=self.was_forced,
+            depth=self.depth,
+            index_in_parent=self.index_in_parent,
+        )
```

Comment:
Does it really make sense to use `was_forced=True` if we're using `with_value=...`?  Or the same `index_in_parent`?

(if so, comments in the code for future readers would be welcome)

Reply:
hmm, it does seem like we would never want to copy+modify a node with `was_forced=True`, at least in the shrinker. Maybe it would be useful elsewhere, but I don't see myself using this combination in the near future. We could `assert not self.was_forced and with_value` for now.

Not sure if this is what you meant by `index_in_parent`, but I'm realizing we should update the location of a node to its new location when using `replacing_nodes=`. Alternatively, we keep copying `(depth, index_in_parent)` like we do now, but anytime we build a tree we have a final `tree.update_locations` which sweeps through and fixes up any invalid locations due to e.g. swapping nodes. I'm reminded of some similar design patterns in [sootup](https://github.com/soot-oss/SootUp).